haskell/test/SMT.hs

```diff
@@ -0,0 +1,808 @@
+import Test.Tasty.QuickCheck
+  ( Arbitrary, Gen, Property, arbitrary, withMaxSuccess, choose, arbitraryBoundedEnum, elements
+  , sublistOf, shuffle, vectorOf)
+import System.Random.SplitMix
+import qualified Text.Show
+import qualified Text.Read
+
+import Tezos.Core (unsafeMkMutez)
+
+import Lorentz (NiceParameter, arg, toVal, fromVal, def)
+import Lorentz.Address
+import Lorentz.Contracts.Stablecoin
+import qualified Lorentz.Contracts.Spec.FA2Interface as FA2
+import Lorentz.Contracts.Test.FA2
+import Lorentz.Contracts.Test.Common
+import Michelson.Test.Integrational
+import Michelson.Test.Dummy
+import Michelson.Interpret
+import Michelson.Typed (untypeValue)
+import Michelson.Text
+import Michelson.TypeCheck
+import qualified Michelson.Typed as T
+import qualified Michelson.Untyped as U
+import Util.Named
+
+-- Some implementation notes:
+--
+-- In this module we test the contract by implementing whole contract
+-- functionality in a Haskell, and comparing its behavior against the actual
+-- contract when run using a Michelson interpretr.
+--
+-- First we generate a random series of (sender address, contract parameter)
+-- pairs. Then we generated a random initial contract state.
+-- Then we apply each of these, both to the Haskell model, and on the
+-- michelson interpretation. After each iteration we check the contract state has not
+-- diverged. If it did, we indicate a test failure. If it did not, we extract the new
+-- contract state and initialize both models with it and apply next parameter, and so on.
+--
+-- For random input generation, we first assign random pool of addresses for
+-- each contract roles.  When we make a contract parameter, say transfer, we
+-- fetch a random sender address from the pool of token owners, and to/from
+-- address (for now we pick operator address and token owner address from the
+-- same pool). This is supposed to create inputs that have a little more
+-- chance of being valid input that has a chance to execute successfully.
+
+-- Right now safelist interaction is not tested, since it seems that doing it
+-- in a useful way require modeling a safelist contract as well.
+
+stablecoinSMTSpec
+  :: Maybe String -- To repeat a failed test, pass the seed printed out by this function here.
+  -> U.Contract
+  -> Spec
+stablecoinSMTSpec mbSeed contracts = do
+  describe "Stable coin contract" $ do
+    it "have the same state as the model after running the inputs" $ do
+        case mbSeed of
+          Just seedStr -> do
+            let seed = QCGen (Text.Read.read seedStr::SMGen)
+            let args = stdArgs{ replay=Just(seed, 0) }
+            result <- quickCheckWithResult args $ withMaxSuccess qcMaxSuccesses $ smtProperty contracts
+            printSeed result
+          Nothing -> do
+            result <- quickCheckResult $ withMaxSuccess qcMaxSuccesses $ smtProperty contracts
+            printSeed result
+    where
+      printSeed result =
```

Comment:
I figured out that the default seed printing doesn't work because you are using `tasty-hspec` and apparently it loses seed printing somehow. I suggest that we just don't use `hspec` at all as it only causes extra indirection and extra problems. So please remove these hacks used to print seed.

Reply:
Removed.